rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+
+### Position-independence of side-loaded tables
+
+We want to ensure that side-loaded tables are maximally re-usable across
+different proofs, even where the 'position' of the same side-loaded table in
+the combined table may vary between proofs. For example, we could consider a
+'folding' circuit `C` which takes 2 side-loaded tables `t_in` and `t_out`,
+where we prove
+```
+C(t_0, t_1)
+C(t_1, t_2)
+C(t_2, t_3)
+...
+```
+and the 'result' of the fold is the final `t_n`. In each pair of executions in
+the sequence, we find that `t_i` is used as both `t_in` and `t_out`, but we
+would like to expose the same table as the value for each.
+
+To achieve this, we use a permutation argument to 'place' the values from each
+side-loaded table at the correct position in the final table. Recall that every
+value
+```
+LookupTable - FixedAndRuntimeTable
+```
+is either 0 (where no side-loaded table contributes a value) or the value of
+some row in `SideLoadedCombined(i) * joint_combiner` from the table `i`.
+
+For the `0` differences, we calculate a permutation between them all and the
+single constant `0`, to ensure that `LookupTable = FixedAndRuntimeTable`. For
+all other values, we set-up a permutation based upon the side-loaded table
+relevant to each value.
+
+Thus, we build the permutation accumulator
+```
+LookupPermutation
+ * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+     + gamma + beta * Sigma(7))
+ * (SideLoadedCombined(0) + gamma + beta * Sigma(8))
+ * (SideLoadedCombined(1) + gamma + beta * Sigma(9))
+ * (SideLoadedCombined(2) + gamma + beta * Sigma(10))
+ * (SideLoadedCombined(3) + gamma + beta * Sigma(11))
+ =
+   LookupPermutation[prev]
+   * ((LookupTable - FixedAndRuntimeTable) * joint_combiner^(-1)
+       + gamma + x * beta * shift[7])
+   * (SideLoadedCombined(0) + gamma + x * beta * shift[8])
+   * (SideLoadedCombined(1) + gamma + x * beta * shift[9])
+   * (SideLoadedCombined(2) + gamma + x * beta * shift[10])
+   * (SideLoadedCombined(3) + gamma + x * beta * shift[11])
+```
+where `shift[0..7]` is the existing `shift` vector used for the witness
+permutation, and the additional values above are chosen so that
+`shift[i] * w^j` are distinct for all `i` and `j`, and `Sigma(_)` are the
+columns representing the permutation. We then assert that
+```
+Lagrange(0)
+  * (LookupPermutation * (0 + gamma + beta * zero_sigma)
+    - (0 + gamma + beta * shift[len(shift)-1]))
+  = 0
+```
+to mix in the contributions for the constant `0` value, and
+```
+Lagrange(n-ZK_ROWS) * (LookupPermutation - 1) = 0
+```
+to ensure that the permuted values cancel.
+
+Note also that the permutation argument allows for interaction **between**
+tables, as well as injection of values into the combined lookup table. As a
+result, the permutation argument can be used to 'copy' parts of one table into
+another, which is likely to be useful in examples like the 'fold' one above, if
+some data is expected to be shared between `t_in` and `t_out`.
```

Comment:
Similar to our main permutation argument which only acts on the first 7 columns, is there a limit on the table columns for which the permutation argument works?  Is there any advantage in doing a similar thing here?